platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalVariantGraphResolveState.java

```diff
@@ -74,22 +75,21 @@ public DefaultLocalVariantGraphResolveState(
     }
 
     @Override
-    public LocalVariantGraphResolveState copyWithTransformedArtifacts(Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> artifactTransformer) {
-
+    public LocalVariantGraphResolveState copyWithComponentId(ComponentIdentifier overrideComponentId) {
         Set<LocalVariantMetadata> artifactSets = artifactState.getArtifactVariants();
         ImmutableSet.Builder<LocalVariantMetadata> copiedArtifactSets = ImmutableSet.builderWithExpectedSize(artifactSets.size());
 
         for (LocalVariantMetadata oldArtifactSet : artifactSets) {
             CalculatedValue<ImmutableList<LocalComponentArtifactMetadata>> newArtifacts =
                 calculatedValueContainerFactory.create(Describables.of(oldArtifactSet.asDescribable(), "artifacts"), c ->
                     oldArtifactSet.getArtifacts().stream()
-                        .map(artifactTransformer::transform)
+                        .map(originalArtifact -> new CompositeProjectComponentArtifactMetadata(overrideComponentId, originalArtifact))
                         .collect(ImmutableList.toImmutableList())
                 );
 
             copiedArtifactSets.add(new LocalVariantMetadata(
                 oldArtifactSet.getName(),
-                oldArtifactSet.getIdentifier(),
+                OverrideComponentIdArtifactSetIdentifier.of(overrideComponentId, oldArtifactSet.getIdentifier()),
```

Comment:
This was the actual bug this PR solves. 
Previously, we use the same identifier for the artifact set, when in reality the artifacts that identifier identified have changed. We need to use a different artifact identifier for this new set.

Reply:
This identifier is used as a key to a cache here https://github.com/gradle/gradle/blob/b6e0160a79046236466eb4c8f3f5ae836d3902a8/platforms/software/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/DefaultVariantArtifactResolver.java#L82